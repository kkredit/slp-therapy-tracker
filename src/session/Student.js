
export const STUDENTS_MIN = 1;
export const STUDENTS_MAX = 4;
export const GOALS_MIN = 1;
export const GOALS_MAX = 4;

export const ATTEMPT_SUCC = 0;
export const ATTEMPT_FAIL = 1;
export const ATTEMPT_CUED = 2;

export class Goal {
  constructor(number) {
    this.key = number;
    this.number = number;
    this.attempts = [];

    this.addAttempt = this.addAttempt.bind(this);
    this.addSuccess = this.addSuccess.bind(this);
    this.addFail = this.addFail.bind(this);
    this.addCued = this.addCued.bind(this);
  }

  addAttempt(type) {
    this.attempts.push(type);
  }

  addSuccess() {
    this.addAttempt(ATTEMPT_SUCC);
  }

  addFail() {
    this.addAttempt(ATTEMPT_FAIL);
  }

  addCued() {
    this.addAttempt(ATTEMPT_CUED);
  }
}

export class Student {
  constructor(number, updateCb) {
    this.key = number;
    this.number = number;
    this.goals = [new Goal(0)];
    this.updateCb = updateCb;

    this.addGoal = this.addGoal.bind(this);
    this.delGoal = this.delGoal.bind(this);
  }

  addGoal() {
    if (this.goals.length < GOALS_MAX) {
      this.goals.push(new Goal(this.goals.length));
      this.updateCb();
    }
  }

  delGoal() {
    if (this.goals.length > GOALS_MIN) {
      this.goals.pop();
      this.updateCb();
    }
  }
}
