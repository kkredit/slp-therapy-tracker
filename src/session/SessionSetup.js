import React from 'react'
import './Session.css'
import { Student, STUDENTS_MIN, STUDENTS_MAX, GOALS_MIN, GOALS_MAX } from './Student.js'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Card from 'react-bootstrap/Card'
import Button from 'react-bootstrap/Button'
import Form from 'react-bootstrap/Form'
import { Link } from 'react-router-dom'

const SetupButton = (props) => {
  return (
    <Button variant="primary" size="sm" className="setupButton" disabled={props.disabled} onClick={props.onClick}>
      <i className={props.faClass} aria-hidden="true"/>
    </Button>
  );
}

const StudentGoalListItem = (props) => {
  return (
    <div>
      Student {props.student.number + 1}:
      <SetupButton faClass="fa fa-minus"
                   disabled={GOALS_MIN >= props.student.goals.length}
                   onClick={props.student.delGoal} />
      {props.student.goals.length}
      <SetupButton faClass="fa fa-plus"
                   disabled={GOALS_MAX <= props.student.goals.length}
                   onClick={props.student.addGoal} />
    </div>
  );
}

const StudentGoalList = (props) => {
  const studentGoalItems = props.students.map((student) => {
    return (
      <StudentGoalListItem key={student.key}
                           student={student} />
    )
  });

  return (
    <React.Fragment>
      {studentGoalItems}
    </React.Fragment>
  );
}

const ProvidersList = (props) => {
  if (!props.providers) {
    return null;
  }
  else {
    const providerListItems = props.providers.map((provider) => {
      return (
        <option key={provider.id}>{provider.fname} {provider.lname}</option>
      )
    });

    return (
      <React.Fragment>
        {providerListItems}
      </React.Fragment>
    );
  }
}

const LocationsList = (props) => {
  if (!props.locations) {
    return null;
  }
  else {
    const locationListItems = props.locations.map((location) => {
      return (
        <option key={location.id}>{location.name}</option>
      )
    });

    return (
      <React.Fragment>
        {locationListItems}
      </React.Fragment>
    );
  }
}

class SessionSetup extends React.Component {
  constructor(props) {
    super(props);

    this.handleInputChange = this.handleInputChange.bind(this);
    this.decrStudentCallback = this.decrStudentCallback.bind(this);
    this.incrStudentCallback = this.incrStudentCallback.bind(this);
    this.studentUpdateCb = this.studentUpdateCb.bind(this);
    this.startSession = this.startSession.bind(this);

    this.state = {
      active: props.active,
      providers: props.providers,
      locations: props.locations,
      provider: null,
      location: null,
      students: [new Student(0, this.studentUpdateCb)],
      studentUpdate: false,
      startCallback: props.startCallback
    };
  }

  static getDerivedStateFromProps(props) {
    return {
      active: props.active,
      providers: props.providers,
      provider: props.providers ? props.providers[0] : null,
      locations: props.locations,
      location: props.locations ? props.locations[0] : null,
    };
  }

  render() {
    if (!this.state.active) {
      return null;
    }
    else {
      return (
        <React.Fragment>
          <Container>
            <Row>
              <Col xs></Col>
              <Col xs={2}>
                <Link className="sideNavButton btn btn-light btn-lg" type="button" to="/history">
                  <i className="fa fa-2x fa-fw fa-angle-double-left" aria-hidden="true"/>
                </Link>
              </Col>
              <Col xs={4}>
                <Card id="setupCard" border="primary" bg="secondary">
                  <Card.Body>
                    <Card.Title> Setup a new session </Card.Title>
                    <Card.Text>
                      Select how many students will be in this session and how many goals each will
                      have.
                    </Card.Text>
                    <Card.Subtitle className="mb-2 setup-hdr"> Provider </Card.Subtitle>
                    <Form.Control as="select" name="provider" onChange={e => this.handleInputChange(e)}>
                      <ProvidersList providers={this.state.providers} />
                    </Form.Control>
                    <Card.Subtitle className="mb-2 setup-hdr"> Location </Card.Subtitle>
                    <Form.Control as="select" name="location" onChange={e => this.handleInputChange(e)}>
                      <LocationsList locations={this.state.locations} />
                    </Form.Control>
                    <Card.Subtitle className="mb-2 setup-hdr"> Number of students </Card.Subtitle>
                    <SetupButton faClass="fa fa-minus"
                                 disabled={STUDENTS_MIN >= this.state.students.length}
                                 onClick={this.decrStudentCallback} />
                    {this.state.students.length}
                    <SetupButton faClass="fa fa-plus"
                                 disabled={STUDENTS_MAX <= this.state.students.length}
                                 onClick={this.incrStudentCallback} />

                    <Card.Subtitle className="mb-2 setup-hdr"> Number of goals </Card.Subtitle>
                    <StudentGoalList students={this.state.students} />
                  </Card.Body>
                </Card>
              </Col>
              <Col xs={2}>
                <Button className="sideNavButton" variant="success" size="lg" onClick={this.startSession}>
                  <i className="fa fa-2x fa-fw fa-angle-right" aria-hidden="true"/>
                </Button>
              </Col>
              <Col xs></Col>
            </Row>
          </Container>
        </React.Fragment>
      );
    }
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    console.log("changing " + name + " to " + value);
    this.setState({
      [name]: value
    });
  }

  decrStudentCallback() {
    if (this.state.students.length > STUDENTS_MIN) {
      this.setState({
        students: this.state.students.slice(0, -1)
      });
    }
  }

  incrStudentCallback() {
    if (this.state.students.length < STUDENTS_MAX) {
      this.setState({
        students: this.state.students.concat(
          [new Student(this.state.students.length, this.studentUpdateCb)]
        )
      });
    }
  }

  studentUpdateCb() {
    this.setState({
      studentUpdate: true
    });
  }

  startSession() {
    this.state.startCallback(this.state.students, this.state.provider, this.state.location);
  }
}

export default SessionSetup;
