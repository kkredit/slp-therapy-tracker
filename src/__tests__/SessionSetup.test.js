import React from 'react';
import ReactDOM from 'react-dom';
import SessionSetup from '../session/SessionSetup.js'
import { Student } from '../session/Student.js'
import GetTestDb from '../__test_helpers__/SetupDb.js'
import { create } from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

it('renders without crashing', () => {
  const div = document.createElement('div');
  let db = GetTestDb();
  ReactDOM.render(
    <BrowserRouter>
      <SessionSetup active={true}
                    startCallback={null}
                    providers={db.providers}
                    locations={db.locations} />
    </BrowserRouter>
  , div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('Snapshots',()=>{
  var tree;
  var instance;

  beforeEach(() => {
    let db = GetTestDb();

    tree = create(
      <BrowserRouter>
        <SessionSetup active={true}
                      startCallback={null}
                      providers={db.providers}
                      locations={db.locations} />
      </BrowserRouter>
    )
    instance = tree.root.findByType(SessionSetup)._fiber.stateNode;
  });

  afterEach(() => {
    tree = null;
    instance = null;
  });

  test('Initial setup', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  })

  test('New student row when have 2 students', () => {
    instance.incrStudentCallback();
    expect(tree.toJSON()).toMatchSnapshot();
  })

  test('Max 4 students', () => {
    instance.incrStudentCallback();
    instance.incrStudentCallback();
    instance.incrStudentCallback();
    instance.incrStudentCallback();
    instance.incrStudentCallback();
    instance.incrStudentCallback();
    expect(tree.toJSON()).toMatchSnapshot();
    expect(instance.state.students.length).toBe(4);
  })
})
