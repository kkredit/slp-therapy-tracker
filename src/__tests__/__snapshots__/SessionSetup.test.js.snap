// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots Initial setup 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    />
    <div
      className="col-2"
    >
      <a
        className="sideNavButton btn btn-light btn-lg"
        href="/history"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-double-left"
        />
      </a>
    </div>
    <div
      className="col-4"
    >
      <div
        className="card bg-secondary border-primary"
        id="setupCard"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
             Setup a new session 
          </div>
          <p
            className="card-text"
          >
            Select how many students will be in this session and how many goals each will have.
          </p>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Provider 
          </div>
          <select
            className="form-control"
            name="provider"
            onChange={[Function]}
          >
            <option>
              sally
               
              foreman
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Location 
          </div>
          <select
            className="form-control"
            name="location"
            onChange={[Function]}
          >
            <option>
              testroom1
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of students 
          </div>
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-minus"
            />
          </button>
          1
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-plus"
            />
          </button>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of goals 
          </div>
          <div>
            Student 
            1
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-2"
    >
      <button
        className="sideNavButton btn btn-success btn-lg"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-right"
        />
      </button>
    </div>
    <div
      className="col"
    />
  </div>
</div>
`;

exports[`Snapshots Max 4 students 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    />
    <div
      className="col-2"
    >
      <a
        className="sideNavButton btn btn-light btn-lg"
        href="/history"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-double-left"
        />
      </a>
    </div>
    <div
      className="col-4"
    >
      <div
        className="card bg-secondary border-primary"
        id="setupCard"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
             Setup a new session 
          </div>
          <p
            className="card-text"
          >
            Select how many students will be in this session and how many goals each will have.
          </p>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Provider 
          </div>
          <select
            className="form-control"
            name="provider"
            onChange={[Function]}
          >
            <option>
              sally
               
              foreman
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Location 
          </div>
          <select
            className="form-control"
            name="location"
            onChange={[Function]}
          >
            <option>
              testroom1
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of students 
          </div>
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-minus"
            />
          </button>
          4
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-plus"
            />
          </button>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of goals 
          </div>
          <div>
            Student 
            1
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
          <div>
            Student 
            2
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
          <div>
            Student 
            3
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
          <div>
            Student 
            4
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-2"
    >
      <button
        className="sideNavButton btn btn-success btn-lg"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-right"
        />
      </button>
    </div>
    <div
      className="col"
    />
  </div>
</div>
`;

exports[`Snapshots New student row when have 2 students 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    />
    <div
      className="col-2"
    >
      <a
        className="sideNavButton btn btn-light btn-lg"
        href="/history"
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-double-left"
        />
      </a>
    </div>
    <div
      className="col-4"
    >
      <div
        className="card bg-secondary border-primary"
        id="setupCard"
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
             Setup a new session 
          </div>
          <p
            className="card-text"
          >
            Select how many students will be in this session and how many goals each will have.
          </p>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Provider 
          </div>
          <select
            className="form-control"
            name="provider"
            onChange={[Function]}
          >
            <option>
              sally
               
              foreman
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Location 
          </div>
          <select
            className="form-control"
            name="location"
            onChange={[Function]}
          >
            <option>
              testroom1
            </option>
          </select>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of students 
          </div>
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-minus"
            />
          </button>
          2
          <button
            className="setupButton btn btn-primary btn-sm"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="fa fa-plus"
            />
          </button>
          <div
            className="mb-2 setup-hdr card-subtitle h6"
          >
             Number of goals 
          </div>
          <div>
            Student 
            1
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
          <div>
            Student 
            2
            :
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-minus"
              />
            </button>
            1
            <button
              className="setupButton btn btn-primary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-plus"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-2"
    >
      <button
        className="sideNavButton btn btn-success btn-lg"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <i
          aria-hidden="true"
          className="fa fa-2x fa-fw fa-angle-right"
        />
      </button>
    </div>
    <div
      className="col"
    />
  </div>
</div>
`;
